src/redirection.rs

```diff
@@ -33,36 +27,21 @@ impl FromStr for Redirection {
     }
 }
 
-#[derive(Debug)]
+#[derive(Debug, Error)]
 pub enum ReadRedirectionError {
-    IOError(io::Error),
-    ParseError(InvalidStrError),
-}
-
-impl Error for ReadRedirectionError {}
-
-impl fmt::Display for ReadRedirectionError {
-    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        match self {
-            Self::IOError(err) => err.fmt(f),
-            Self::ParseError(err) => err.fmt(f),
-        }
-    }
+    #[error("IO operation failed")]
+    IOError(#[from] io::Error),
+    #[error(transparent)]
+    ParseError(#[from] InvalidStrError),
 }
```

Comment:
this is the magic of `thiserror`.

it implements the `Display` for you via `#[error("my display)]`

it can also implement `impl From<T> for MyError {..}` so that you can do `error?` easier